driver-core/src/main/java/com/datastax/driver/core/PercentileTracker.java

```diff
@@ -0,0 +1,339 @@
+            return new CachedHistogram(Futures.<Histogram>immediateFuture(null));
+        }
+    }
+
+    @Override
+    public void onRegister(Cluster cluster) {
+        // nothing by default
+    }
+
+    @Override
+    public void onUnregister(Cluster cluster) {
+        // nothing by default
+    }
+
+    /**
+     * Determines whether a particular measurement should be included.
+     * <p/>
+     * This is used to ignore measurements that could skew the statistics; for example, we typically want to ignore
+     * invalid query errors because they have a very low latency and would make a given cluster/host appear faster than
+     * it really is.
+     *
+     * @param host      the host that was queried.
+     * @param statement the statement that was executed.
+     * @param exception if the query failed, the corresponding exception.
+     * @return whether the measurement should be included.
+     */
+    protected boolean include(Host host, Statement statement, Exception exception) {
+        // query was successful: always consider
+        if (exception == null)
+            return true;
+        // filter out "fast" errors
+        // TODO this was copy/pasted from LatencyAwarePolicy, maybe it could be refactored as a shared method
+        return !EXCLUDED_EXCEPTIONS.contains(exception.getClass());
+    }
+
+    /**
+     * A set of DriverException subclasses that we should prevent from updating the host's score.
+     * The intent behind it is to filter out "fast" errors: when a host replies with such errors,
+     * it usually does so very quickly, because it did not involve any actual
+     * coordination work. Such errors are not good indicators of the host's responsiveness,
+     * and tend to make the host's score look better than it actually is.
+     */
+    private static final Set<Class<? extends Exception>> EXCLUDED_EXCEPTIONS = ImmutableSet.<Class<? extends Exception>>of(
+            UnavailableException.class, // this is done via the snitch and is usually very fast
+            OverloadedException.class,
+            BootstrappingException.class,
+            UnpreparedException.class,
+            QueryValidationException.class // query validation also happens at early stages in the coordinator
+    );
+
+    /**
+     * Base class for {@code PercentileTracker} implementation builders.
+     *
+     * @param <B> the type of the concrete builder implementation.
+     * @param <T> the type of the object to build.
+     */
+    public static abstract class Builder<B, T> {
+        protected final long highestTrackableLatencyMillis;
+        protected int numberOfSignificantValueDigits = 3;
+        protected int minRecordedValues = 1000;
+        protected long intervalMs = MINUTES.toMillis(5);
+
+        Builder(long highestTrackableLatencyMillis) {
+            this.highestTrackableLatencyMillis = highestTrackableLatencyMillis;
+        }
+
+        protected abstract B self();
```

Comment:
Nice!  I've never seen Builders generalized so they may be extended by a subclass, usually there is one for each implementation. The self trick is cool.